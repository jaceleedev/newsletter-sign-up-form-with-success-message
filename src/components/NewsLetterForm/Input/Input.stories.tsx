import type { Meta, StoryObj } from '@storybook/react';

import Input from './Input';

const meta = {
  component: Input,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    id: {
      description: 'Unique identifier to link the label',
      control: 'text',
    },
    label: {
      description: 'Text label to be displayed above the input field',
      control: 'text',
    },
    isValid: {
      description: 'Boolean indicating whether the input value is invalid',
      control: 'boolean',
    },
    errorMessage: {
      description: 'Message to be displayed when an error occurs',
      control: 'text',
    },
  },
  args: {
    type: 'email',
  },
  decorators: [
    (Story) => (
      <div style={{ width: '327px' }}>
        <Story />
      </div>
    ),
  ],
} satisfies Meta<typeof Input>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  decorators: [
    (Story) => (
      <div style={{ width: '327px' }}>
        <Story />
      </div>
    ),
  ],
  args: {
    id: 'email',
    label: 'Email address',
    errorMessage: 'Valid email required',
    isValid: true,
    placeholder: 'email@company.com',
  },
};

export const Focused: Story = {
  decorators: [
    (Story) => (
      <div style={{ width: '327px' }}>
        <Story />
      </div>
    ),
  ],
  args: {
    id: 'email',
    label: 'Email address',
    isValid: true,
    errorMessage: 'Valid email required',
    value: 'ash@loremcompany.com',
  },
  parameters: {
    pseudo: { focus: true },
  },
};

export const Invalid: Story = {
  args: {
    id: 'email',
    label: 'Email address',
    isValid: false,
    errorMessage: 'Valid email required',
    value: 'ash#loremcompany.com',
  },
};
