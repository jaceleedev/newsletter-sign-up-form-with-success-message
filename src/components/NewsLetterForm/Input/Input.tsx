import React, { InputHTMLAttributes, useMemo } from 'react';

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  id: string;
  label: string;
  isValid: boolean;
  errorMessage: string;
}

const BASE_STYLES = [
  'text-body',
  'w-full',
  'h-14',
  'pl-6',
  'py-4',
  'rounded-lg',
  'border',
  'border-solid',
  'outline-none',
  'text-dark-navy',
  'focus:border-dark-navy',
  'placeholder:opacity-50',
  'transition-all',
];

const INVALID_STYLES = [
  'text-vermellion',
  'border-vermellion',
  'bg-vermellion',
  'focus:text-vermellion',
  'focus:border-vermellion',
  'bg-opacity-15',
];

const VALID_STYLES = ['border-[rgba(25,24,43,0.25)]', 'bg-white'];

function generateInputStyles(isValid: boolean) {
  return [...BASE_STYLES, ...(isValid ? VALID_STYLES : INVALID_STYLES)].join(
    ' '
  );
}

function Input({
  id,
  label,
  errorMessage,
  isValid = true,
  ...props
}: InputProps) {
  const computedStyles = useMemo(() => generateInputStyles(isValid), [isValid]);

  return (
    <div className="grid grid-cols-1 gap-2 w-full">
      <div className="flex justify-between items-center">
        <label htmlFor={id} className="text-body-small text-dark-navy">
          {label}
        </label>
        {!isValid && (
          <span
            className="text-body-small text-vermellion"
            role="alert"
            aria-live="polite"
          >
            {errorMessage}
          </span>
        )}
      </div>
      <input
        id={id}
        type={props.type || 'text'}
        className={computedStyles}
        aria-invalid={!isValid}
        aria-required={props.required}
        {...props}
      />
    </div>
  );
}

export default React.memo(Input);
