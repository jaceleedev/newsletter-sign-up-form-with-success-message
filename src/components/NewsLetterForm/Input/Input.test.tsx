import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import Input from './Input';

expect.extend(toHaveNoViolations);

describe('Input', () => {
  // 1. Rendering test
  it('renders correctly', () => {
    render(
      <Input
        id="rendering-test"
        label="Rendering test"
        isValid={true}
        errorMessage="Rendering message"
      />
    );
    const input = screen.getByLabelText('Rendering test');
    expect(input).toBeDefined();
  });

  // 2. Props test
  it('applies correct attributes', () => {
    render(
      <Input
        id="props-test"
        label="Props test"
        type="email"
        isValid={true}
        errorMessage="Props test"
        placeholder="Props message"
      />
    );
    const input = screen.getByLabelText('Props test');
    expect(input).toHaveProperty('id');
    expect(input).toHaveProperty('type');
    expect(input).toHaveProperty('placeholder');
  });

  // 3. Styles test
  it('applies the correct styles', () => {
    render(
      <Input
        id="styles-test"
        label="Styles test"
        isValid={true}
        errorMessage="Styles message"
      />
    );
    const input = screen.getByLabelText('Styles test');
    expect(input.className).toContain('text-body');
    expect(input.className).toContain('w-full');
    expect(input.className).toContain('h-14');
    expect(input.className).toContain('rounded-lg');
  });

  // 4. error styles test
  it('applies invalid styles when isValid prop is false', () => {
    render(
      <Input
        id="error-styles-test"
        label="Error styles test"
        isValid={false}
        errorMessage="Error styles message"
      />
    );
    const input = screen.getByLabelText('Error styles test');
    expect(input.className).toContain('focus:text-vermellion');
    expect(input.className).toContain('focus:border-vermellion');
    expect(input.className).toContain('bg-vermellion');
    expect(screen.getByText('Error styles message')).toBeDefined();
  });

  // 4. State change test
  it('shows error message when invalid', () => {
    render(
      <Input
        id="invalid-test"
        label="Invalid test"
        isValid={false}
        errorMessage="Invalid message"
      />
    );
    const errorMessage = screen.getByText('Invalid message');
    expect(errorMessage).toBeDefined();
    expect(errorMessage).toHaveProperty('role');
  });

  // 5. Behavior test
  it('calls onChange when input value changes', () => {
    const handleChange = vi.fn();
    render(
      <Input
        id="behavior-test"
        label="Behavior test"
        isValid={true}
        errorMessage="Behavior message"
        onChange={handleChange}
      />
    );
    const input = screen.getByLabelText('Behavior test');
    fireEvent.change(input, { target: { value: 'test' } });
    expect(handleChange).toHaveBeenCalledOnce();
  });

  // 6. Accessibility test
  it('has no accessibility violations', async () => {
    const { container } = render(
      <Input
        id="accessibility"
        label="Accessibility test"
        isValid={true}
        errorMessage="Accessibility message"
      />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  // 7. Snapshot test
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Input
        id="snapshot"
        label="Snapshot test"
        isValid={true}
        errorMessage="Snapshot message"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
