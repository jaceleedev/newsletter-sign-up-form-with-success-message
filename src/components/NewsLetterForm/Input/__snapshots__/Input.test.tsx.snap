// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-1 gap-2 w-full"
  >
    <div
      class="flex justify-between items-center"
    >
      <label
        class="text-body-small text-dark-navy"
        for="snapshot"
      >
        Snapshot test
      </label>
    </div>
    <input
      aria-invalid="false"
      class="text-body w-full h-14 pl-6 py-4 rounded-lg border border-solid outline-none text-dark-navy focus:border-dark-navy placeholder:opacity-50 transition-all border-[rgba(25,24,43,0.25)] bg-white"
      id="snapshot"
      type="text"
    />
  </div>
</DocumentFragment>
`;
